Doc:Tag tracing existing Public API objects

marcotc opened the pull request:
This PR adds the `@public_api` tag, introduced in #1809, to existing objects in `ddtrace`'s core and "tracing" product that will now be under strict semver come 1.0.

Our docs look this now:


![Screen Shot 2021-12-21 at 5 28 07 PM](https://user-images.githubusercontent.com/583503/147019371-187040d1-7f24-4b44-9aae-2dda5692b211.png)



Notably, this PR does not include CI-App and Profiling API tagging. These will be done in their own PRs.

Also, the new main [`Datadog::Tracing` public API module](https://github.com/DataDog/dd-trace-rb/pull/1778/files#diff-5d787c2112f23c092e017ace55b113a3cd5098404246b3d1c5434f6c27602915) is also missing, as it is an entirely new module, and thus will be introduced in a separate PR, along with its respective test suite.

This PR adds a new tag `@default default_value_here`, which creates a section for attribute default values. This is useful for settings documentation: 
![Screen Shot 2021-12-13 at 1 35 16 PM](https://user-images.githubusercontent.com/583503/145892579-cf6d1e95-e816-479a-a641-b582454a7b99.png)

This PR also expands the documentation on any public API methods that I was confident enough to tackle :)

## Functional changes

There are a couple non-doc changes
1. The default value for the `partial_flush.min_spans_threshold` setting has been changed from `nil` to `500`. This effectively causes no changes as `500` is the default when `nil` is provided in the partial flusher. This change only exposes the default to the user, as opposed to hiding it behind an implementation detail.
2. The SyncWriter initializer now accepts formal keyword arguments, instead of a generic `options = {}`.

ivoanjo commented:
I get this weird effect where sometimes I don't get the class list on the right column, but I can't quite reproduce it (was not working a few hours ago -- now it's working).

I'll keep an eye out for it, decided to drop a note just in case you saw it as well.

(I'm working on the actual review part :D -- 🔜 )

ivoanjo commented on lib/datadog/core/environment/ext.rb:
```diff
@@ -4,6 +4,7 @@
 module Datadog
   module Core
     module Environment
+      # @public_api
```
A bunch of these modules aren't actually showing up on the sidebar:

<img width="256" alt="Screen Shot 2021-12-14 at 19 00 26" src="https://user-images.githubusercontent.com/2785847/146054119-6f0ae878-0200-4fab-b007-7f8e730d2e1f.png">

-- is this deliberate?

marcotc replied:
Addresses this!

ivoanjo commented on lib/datadog/core/environment/ext.rb:
```diff
@@ -4,6 +4,7 @@
   module Core
     module Environment
+      # @public_api
       module Ext
```
Does this need to be public? 🤔 

It seems a bit limiting for this to be a public API, especially because customers can trivially get all of this from the same place we get 'em.

Also applies to other similar classes/modules such as Environment::Identity, Environment::VariableHelpers, etc

ivoanjo commented on lib/ddtrace.rb:
```diff
@@ -23,7 +23,8 @@
 require 'ddtrace/opentelemetry/extensions'
 
-# \Datadog global namespace that includes all tracing functionality for Tracer and Span classes.
+# {Datadog} global namespace that includes all tracing functionality for Tracer and Span classes.
```
```suggestion
# Global namespace that includes all Datadog functionality.
```

This description is somehow outdated :)

ivoanjo commented on lib/ddtrace/auto_instrument.rb:
```diff
@@ -1,4 +1,7 @@
 # typed: strict
+# Entrypoint file for auto instrumentation.
+#
+# This file's path is part of the @public_api.
```
May be worth adding that the public API here is the path to this file, as it doesn't otherwise expose any other user-callable code.

marcotc replied:
> May be worth adding that the public API here is the path to this file, as it doesn't otherwise expose any other user-callable code.

This is what I tried to convey with `This file's path is part of the @public_api.`. Do you think that's not clear?

ivoanjo replied:
Uhhh 😰 I somehow misread this, I guess I read it as "This file is part of the @public_api" or something. Please ignore >_>

marcotc commented:
> I get this weird effect where sometimes I don't get the class list on the right column, but I can't quite reproduce it

Does it get "fixed" when you generated the documentation again, or when you refresh the browser?

I noticed that YARD doesn't really bust locally generated files in between generation runs, and I've stumbled upon inconsistent docs many times, having to nuke the directory a few times.

ivoanjo commented on lib/ddtrace/configuration/base.rb:
```diff
@@ -5,6 +5,7 @@
 module Datadog
   module Configuration
     # Basic configuration behavior
+    # @public_api
     module Base
```
I'm curious about the `@public_api` declarations on this file as well, since customers don't use this directly... (I think...?)

ivoanjo commented on lib/ddtrace/configuration/option.rb:
```diff
@@ -2,6 +2,7 @@
 module Datadog
   module Configuration
     # Represents an instance of an integration configuration option
+    # @public_api
     class Option
```
Usual public question -- this seems to be the option system implementation; why public?

marcotc replied:
Because `Option`'s public methods are used when uses create a configuration DSL for their custom instrumentations.

ivoanjo replied:
Hmm... but do customers ever see these methods?

My thinking is that I'm looking at [`settings.rb`](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/configuration/settings.rb) and even though those things get turned into `Options` instances -- we never actually see the `Options` class or any of its methods being called or referenced there. This was what led me to question this (and a few others) -- I understand that they get used, but it seems to me that they are the implementation details of mechanisms, and that customers, even advanced ones, are probably not going to need to know about them directly.

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -42,19 +44,44 @@ def initialize(*_)
         end
       end
 
+      # Profiler API key.
+      # @default `DD_API_KEY` environment variable, otherwise `nil`
+      # @return [String,nil]
+      # @public_api
       option :api_key do |o|
```
This one is a though one, since it's not really just for the profiler (other datadog producs use `DD_API_KEY`), it's just that only the profiler /right now/ makes use of it.

I suggest marking it as "For internal use only", even if it's still public (maybe...?).

marcotc replied:
Should we have the option, but not make it part of the public API then, given your rationale?
I don't think I grasp 100% the situation here.

ivoanjo replied:
> I don't think I grasp 100% the situation here.

Yup, sorry for the confusion.

My thinking is -- we want to keep these options. With that in mind, I think it doesn't quite make sense to have "private options". But documentation-wise, we still want to steer customers away from them.

Something like this (lol):

![](https://previews.123rf.com/images/dondesigns/dondesigns1409/dondesigns140900005/31501377-do-not-press-push-button-isolated-on-white-background.jpg)

So my proposal is that for site and api_key we:

1. Don't mark it as `@public_api`
2. Don't explain what it is (or at least, an explanation of what it is should not appear in the generated docs -- an internal comment is fine)
3. Just state that this is "For internal Datadog testing use. Changing it is not recommended".

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -42,19 +44,44 @@ def initialize(*_)
         o.lazy
       end
 
+      # Internal tracer diagnostic settings.
+      #
+      # Enabling these surfaces debug information that can be helpful to
+      # diagnose issues related to the tracer internals.
+      # @public_api
       settings :diagnostics do
+        # Outputs all spans created by the host application to `Datadog.logger`.
+        #
+        # **This option is very verbose!** It's only recommended for non-production
+        # environments.
+        #
+        # This option is helpful when trying to understand what information the
+        # tracer is sending to the Agent or backend.
+        # @default `DD_TRACE_DEBUG` environment variable, otherwise `false`
+        # @return [Boolean]
         option :debug do |o|
```
This actually enables debug logging on the default logger, so it applies to all products, not just tracer.

(I wish we had a way to enable some debug logging without the full trace output, since it gets really noisy while debugging >_>)

marcotc replied:
>(I wish we had a way to enable some debug logging without the full trace output, since it gets really noisy while debugging >_>)

Debug levels? :)

Or we can add a `span_debug` (wip name) option that controls only span output. I agree that span output is much more verbose than anything else.

I guess it would be similar to the profiler outputting all its collected stacks.

ivoanjo replied:
Log4j jokes aside, I've spent more time even in the Java ecosystem than in the Ruby ecosystem, and I think Java logging frameworks solve it in quite a reasonable way -- by creating a new logger instance per class:

```java

Logger logger = Logger.getLogger(MyClass.class.getName());
```

Then, most logging frameworks allow you to configure per-class (or even per namespace) logging levels, so solving this "issue" in would be as simple as enabling debug logging for everything other than the class that prints out the full spans.

The default Ruby logger does not have this behavior, but there are some third party logger gems that do, for instance [logging](https://github.com/twp/logging) and [semantic logger](https://logger.rocketjob.io/).

So I'd actually lean into either using a better 3rd-party logger or enhancing the one we already use to get these kinds of behaviors, rather than having ad-hoc options for some of the more noisy things.

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -65,16 +92,38 @@ def initialize(*_)
           end
         end
 
+        # Internal tracer {Datadog::Statsd} metrics collection.
+        #
+        # The list of metrics collected can be found in {Datadog::Ext::Diagnostics::Health::Metrics}.
         settings :health_metrics do
```
Minor: Should we not mention the tracer here, since this is in a generic group and other products may end up using it as well?

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -65,16 +92,38 @@ def initialize(*_)
           option :statsd
         end
 
+        # Tracer startup debug log statement configuration.
         settings :startup_logs do
+          # Enable startup logs collection.
+          #
+          # If `nil`, defaults to logging startup logs when `ddtrace` detects that the application
+          # is *not* running in a development environment.
+          #
+          # @default `DD_TRACE_STARTUP_LOGS` environment variable, otherwise `nil`
+          # @return [Boolean,nil]
           option :enabled do |o|
```
Probably one more that could be made generic, although it does somewhat say trace on the env flag :/

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -198,6 +314,9 @@ def runtime_metrics(options = nil)
       end
 
+      # TODO: remove me. Use `c.runtime_metrics.enabled =`
+      # @deprecated Use `runtime_metrics.enabled` instead.
+      # @return [Boolean]
       option :runtime_metrics_enabled do |o|
```
Minor: May be worth stating here that this is deprecated for removal (see also [how Java does it nowadays](https://docs.oracle.com/javase/9/core/enhanced-deprecation1.htm#JSCOR-GUID-E182DAF6-282B-4C1F-8CCF-0A61008E3F11))

marcotc replied:
I'm thinking that something marked as deprecated can be removed in the next major version unceremoniously.

What do you think?

(I do like how explicit `@Deprecated(since="9", forRemoval=true)` reads!)

ivoanjo replied:
> I'm thinking that something marked as deprecated can be removed in the next major version unceremoniously.

I'm happy with that definition, but my concern is that it may not be clear to customers.

It's a bit like that for Java as well -- some APIs have been deprecated for 10+ (20+?) years and they're still there. So I think it's important to clarify that we mean the "it's deprecated because we're going to remove it soon" and not the "it's deprecated because you shouldn't use it, but well, we're going to keep it here".

Perhaps we could add a small text block explaining what deprecation means in our library and link it whenever we mark a method as deprecated? Something like

```
# @deprecated [For removal](...nice link so we don't repeat ourselves...), use `runtime_metrics.enabled` instead.
```

marcotc replied:
Done this suggestion: 
![Screen Shot 2021-12-21 at 2 13 19 PM](https://user-images.githubusercontent.com/583503/147004464-1087977b-fc58-4c52-a897-820590726fd3.png)

Which links to: 
![Screen Shot 2021-12-21 at 2 13 23 PM](https://user-images.githubusercontent.com/583503/147004475-1a534ca2-3c2b-44e8-ac12-e6dd4ded641c.png)

ivoanjo replied:
Seems great, thanks!

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -232,11 +373,20 @@ def runtime_metrics(options = nil)
         end
       end
 
+      # TODO: profiler related. Ask @ivoanjo what this is for.
+      # @return [String,nil]
+      # @public_api
       option :site do |o|
         o.default { ENV.fetch(Ext::Environment::ENV_SITE, nil) }
         o.lazy
       end
```
This is paired with `DD_API_KEY` when talking to the backend -- it's the actual deployment (us, eu, ...) [like here](https://docs.datadoghq.com/agent/troubleshooting/site/). My suggestion is to mark it in the same way as we do the api key -- by stating it's for internal Datadog use only.

marcotc replied:
Thank you for that link above! Super helpful!

Documented as such.

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -232,11 +373,20 @@ def runtime_metrics(options = nil)
         o.lazy
       end
 
+      # Default tracing span tags.
+      #
+      # These tags are applied to every span.
+      # @default `DD_TAGS` environment variable (in the format `'tag1:value1,tag2:value2'`), otherwise `{}`
+      # @return [Hash<String,String>]
+      # @public_api
       option :tags do |o|
```
These are also the default tags for the profiler ;)

ivoanjo commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -277,23 +427,45 @@ def runtime_metrics(options = nil)
           o.lazy
         end
 
+        # TODO: Remove this configuration.
+        # TODO: It is not necessary, as it be configured by the default flushing trace configuration.
         option :trace_flush do |o|
           o.default { nil }
           o.lazy
         end
 
+        # TODO: Remove this configuration.
+        # TODO: It is not necessary, as it be configured by the default writer trace configuration.
         option :writer_options do |o|
```
Are these TODOs going to be visible to customers? If so, should we mark them as deprecated instead?

marcotc replied:
YARD is naive, it keeps everything that is a comment before the declaration, TODOs and all.
I'll see if I can clean them up.

marcotc replied:
TODOs and DEVs don't surface anymore.

ivoanjo reviewed (commented):
Wow this was more than I was expecting 😅. Here's what I got so far, will keep on in tomorrow.

ivoanjo commented:
> Does it get "fixed" when you generated the documentation again, or when you refresh the browser?
>
> I noticed that YARD doesn't really bust locally generated files in between doc generation runs, and I've stumbled upon inconsistent docs many times, having to nuke the directory a few times.
>
> The official docs shouldn't be affected by it, if that's the case your are seeing, but if it's on browser refresh then there's more to it.

Yeah I think that was it. I'll keep an eye out for it :)

ivoanjo commented on lib/ddtrace/sampler.rb:
```diff
@@ -6,22 +6,44 @@
 
   # \AllSampler samples all the traces.
+  # @public_api
   class AllSampler < Sampler
```
Is it just me that finds it weird that AllSampler is at the top level at `Datadog::AllSampler`? Seems kinda a prominent location for such trivial details.

marcotc replied:
It's kind of weird.

ivoanjo commented on lib/ddtrace/span_operation.rb:
```diff
@@ -15,6 +15,7 @@ module Datadog
   # build a Span. When completed, it yields the Span.
   #
   # rubocop:disable Metrics/ClassLength
```
⬆️ The rubocop magic comment is showing up in the generated docs (same for `TraceOperation` and `Tracer`)

marcotc replied:
> The rubocop magic comment is showing up in the generated docs

I feel like this is a common enough thing that YARD should have some support for it out of the box. I noticed this during development, but didn't address each one, because there was no way 🙃.

I couldn't find anything regarding filtering out docstrings, so I'm thinking of proposing something upstream.

In the meantime, I'll write an extension to remove rubocop/magic strings.

marcotc replied:
These should have been removed in the latest commit.

ivoanjo commented on lib/ddtrace/tracer.rb:
```diff
@@ -215,6 +257,7 @@ def publish(trace)
     #
     #   tracer.shutdown!
```
⬆️ This and a few more methods on this class have code examples that are outside ```` ``` ```` blocks and thus don't get monospaced/syntax highlighted.

marcotc replied:
Fixed two cases of examples outside of a code block in this file.

ivoanjo reviewed (commented):
Ok... I have reviewed all 212 files 😓

I still wish that we went with smaller PRs for these big 1.0 refactoring changes. I know we need to do all of them, but it's taking me multiple hours to go through each one :(

marcotc commented:
> I still wish that we went with smaller PRs for these big 1.0 refactoring changes. I know we need to do all of them, but it's taking me multiple hours to go through each one :(

I'm not sure how to best divide this PR. I could break it down by folder/module if you think that's easier.

marcotc commented:
The hard part is that I documented a bunch of stuff that was never documented before, so there's like 4 year old methods being documented just now, and I kept finding underdocumented stuff as the work went on...

ivoanjo commented:
> The hard part is that I documented a bunch of stuff that was never documented before, so there's like 4 year old methods being documented just now, and I kept finding underdocumented stuff as the work went on...

My suggestion would be to slice it by large components, for instance core/tracing/open[telemetry+tracing]/integrations.

I realize that when working on it that may not have been the order (or it may be harder to work in that way), but I still think after-the-fact breaking it into smaller units is 100% worth it, because otherwise it's really easy to just gloss over such a large PR and not give it the attention it deserves.

marcotc commented on lib/ddtrace/opentelemetry/extensions.rb:
```diff
@@ -5,6 +5,7 @@
 module Datadog
   module OpenTelemetry
     # Defines extensions to ddtrace for OpenTelemetry support
+    # @public_api
     module Extensions
       def self.extended(base)
         Datadog::SpanOperation.prepend(OpenTelemetry::Span)
```
I was going to check exactly why it's public, but we've just had an internal discussion today, and it seems like we are moving in a different direction with Otel.
I'm going to remove these public api markings for now.

marcotc commented on lib/ddtrace/configuration/settings.rb:
```diff
@@ -310,27 +482,82 @@ def runtime_metrics(options = nil)
           o.default { env_to_bool(Datadog::Ext::Diagnostics::DD_TRACE_ENABLED, true) }
           o.lazy
         end
         option :hostname # TODO: Deprecate
```
I expanded on the ones I understand (hostname and port). The others I'm not sure. I'll sync with @delner soon to cross reference with him.

marcotc commented on lib/ddtrace/contrib/action_cable/ext.rb:
```diff
@@ -3,6 +3,7 @@ module Datadog
   module Contrib
     module ActionCable
       # ActionCable integration constants
+      # @public_api
       module Ext
```
This is here to protect the resource name values and tag name values, as changing those is a pretty big breaking change.
I'll add a comment to the integrations to explain that.

marcotc commented on lib/ddtrace/contrib/action_cable/integration.rb:
```diff
@@ -13,6 +13,7 @@ class Integration
 
         MINIMUM_VERSION = Gem::Version.new('5.0.0')
 
+        # @public_api
         register_as :action_cable, auto_patch: false
```
I added here as a documentation to us, I'm not sure how to best capture this: changes to this `register_as` line as normally large breaking changes (either renaming an integration or changing its auto_patch behaviour).

marcotc replied:
Barred a better solution, I added a better rationale on each `@public_api` tagging of `register_as`.

marcotc commented on lib/ddtrace/contrib/analytics.rb:
```diff
@@ -4,6 +4,7 @@
 module Datadog
   module Contrib
     # Defines analytics behavior for integrations
+    # @public_api
     module Analytics
       module_function
 
```
ₛₘₐₗₗ convenience module :hide:
I agree that it doesn't seem necessary.

marcotc commented on lib/ddtrace/contrib/configurable.rb:
```diff
@@ -31,6 +31,7 @@ module InstanceMethods
         # DEV(1.0): Unfortunately, change this would be a breaking change for all custom integrations,
         # DEV(1.0): thus we have to be very intentional with the right time to make this change.
         # DEV(1.0): Currently marking this for a 1.0 milestone.
+        # @public_api
```
Alright, let's make this breaking change then before 1.0.
I removed the public tag, thank you!

marcotc commented on lib/ddtrace/contrib/extensions.rb:
```diff
@@ -87,6 +87,7 @@ module Settings
           # @param [Symbol] integration_name the integration name
           # @param [Object] key the integration-specific lookup key
           # @return [Datadog::Contrib::Configuration::Settings]
+          # @public_api
           def [](integration_name, key = :default)
```
Documented inline now: `# @public_api Used for Datadog.configuration[:my_integration]`

marcotc commented on lib/ddtrace/contrib/patchable.rb:
```diff
@@ -9,7 +9,11 @@ def self.included(base)
+        # This is the gem version, when the instrumentation target is a Ruby gem.
         def version
           nil
         end
```
I've re-written and expanded all the docs in this file. Let me know if this makes more sense now.

ivoanjo replied:
Thanks, I really like the customer focus on the rewritten docs!

marcotc commented on lib/ddtrace/contrib/patchable.rb:
```diff
@@ -30,12 +34,16 @@ def compatible?
         end
```
That seems reasonable to me 👍 
I'll add it to the list of things to do before 1.0.

marcotc commented on lib/ddtrace/contrib/registerable.rb:
```diff
@@ -10,12 +10,25 @@ def self.included(base)
+        # @param [Datadog::Contrib::Registry] registry a custom registry. Defaults to the global tracing registry.
+        # @param [Boolean] auto_patch will this integration be activated during
+        #   {file:docs/AutoInstrumentation.md Auto Instrumentation}?
+        # @param [Hash] options additional keyword options passed to the initializer of
+        #   a custom {Registerable} instrumentation
```
> Are these options that can be used by a custom integration?

Correct.
I think it's one of those things that you'd see it's use if you were writing a custom instrumentation.

It's the `auto_patch: true` option, for example.

marcotc commented on lib/ddtrace/contrib/registerable.rb:
```diff
@@ -10,12 +10,25 @@ def self.included(base)
+        #
+        # @param [Symbol] name integration name. Used during activation.
+        # @param [Datadog::Contrib::Registry] registry a custom registry. Defaults to the global tracing registry.
+        # @param [Boolean] auto_patch will this integration be activated during
+        #   {file:docs/AutoInstrumentation.md Auto Instrumentation}?
```
Great idea, done!

marcotc commented on lib/ddtrace/contrib/status_code_matcher.rb:
```diff
@@ -5,6 +5,7 @@
 module Datadog
   module Contrib
     # Contains methods helpful for tracing/annotating HTTP request libraries
+    # @public_api
     class StatusCodeMatcher
```
👍

marcotc commented on lib/ddtrace/correlation.rb:
```diff
@@ -8,6 +8,7 @@ module Datadog
+      # @public_api
       attr_reader \
         :env,
         :service,
```
Seems reasonable.
Also, the visibility on this class was kind of botched, so I redid it just now as well.

marcotc commented on lib/ddtrace/opentracer/binary_propagator.rb:
```diff
@@ -2,13 +2,15 @@
 module Datadog
   module OpenTracer
     # OpenTracing propagator for Datadog::OpenTracer::Tracer
+    # @public_api
     module BinaryPropagator
```
👍

marcotc commented on lib/ddtrace/opentracer/tracer.rb:
```diff
@@ -4,52 +4,55 @@
       attr_reader \
         :datadog_tracer
 
-      def_delegators \
-        :datadog_tracer,
-        :configure
```
The `#configure` method was removed in https://github.com/DataDog/dd-trace-rb/pull/1753. I forgot to clean up this delegator, that now goes nowhere.

marcotc commented on lib/ddtrace/propagation/grpc_propagator.rb:
```diff
@@ -8,6 +8,7 @@ module Datadog
   # to the HTTPPropagator; the key difference is the way gRPC handles
   # header information (called "metadata") as it operates over HTTP2
+  # @public_api
   module GRPCPropagator
```
I think I was being "consistent" with having `HTTPPropagator` exposed, but `HTTPPropagator` is used commonly when a custom HTTP framework is used by users, which is not rare.

But a custom gRPC implementation seems unlikely, I've hidden it.

marcotc commented on lib/ddtrace/tracer.rb:
```diff
@@ -65,44 +65,57 @@ def initialize(
-    # * +type+: the type of the span (such as \http, \db and so on)
+    # @param [String] name {Datadog::Span} operation name.
+    #   See {https://docs.datadoghq.com/tracing/guide/configuring-primary-operation/ Primary Operations in Services}.
+    # @param [Datadog::TraceDigest] continue_from continue a trace from a {Datadog::TraceDigest}. Used for async.
```
Expanded that part.

marcotc commented on lib/ddtrace/tracer.rb:
```diff
@@ -65,44 +65,57 @@ def initialize(
+    # @return [Datadog::SpanOperation] If no block is provided, returns the active, unfinished {Datadog::SpanOperation}.
+    # @yield Optional block where new newly created {Datadog::SpanOperation} captures the execution.
+    #   The span_op is {Datadog::SpanOperation#finish}ed when the block ends.
+    #   If no block, the active, unfinished {Datadog::SpanOperation} is returned instead.
```
Agree, removed redundant part.

marcotc commented on lib/ddtrace/tracer.rb:
```diff
@@ -151,23 +164,41 @@ def trace(
     # A valid example is:
     #
     #   tracer.set_tags('env' => 'prod', 'component' => 'core')
+    # @public_api
     def set_tags(tags)
       string_tags = tags.collect { |k, v| [k.to_s, v] }.to_h
       @tags = @tags.merge(string_tags)
     end
 
-    # Return the current active trace or +nil+.
+    # The active, unfinished trace, representing the current instrumentation context.
+    #
+    # The active trace is thread-local.
+    #
+    # @return [Datadog::TraceSegment] the active trace
+    # @return [nil] if no trace is active
+    # @public_api
     def active_trace(key = nil)
```
I want to hide `key`, let me see if there's a good way to do it.

marcotc replied:
No dice, I just expanded the documented and added "For internal use only.".

marcotc commented on lib/ddtrace/tracer.rb:
```diff
@@ -176,14 +207,25 @@ def active_correlation(key = nil)
       call_context(key).activate!(trace, &block)
     end
 
+    # TODO: make this private
     def trace_completed
```
Agreed, hid it.

marcotc commented on lib/ddtrace/writer.rb:
```diff
@@ -17,6 +17,7 @@ class Writer
       :worker,
       :events
 
+    # @public_api
     def initialize(options = {})
```
I didn't tackle it because the future is `Datadog::Workers::TraceWriter` instead.

I added an internal `DEV:` comment to this class to document this.

marcotc commented on lib/ddtrace/writer.rb:
```diff
@@ -49,6 +50,11 @@ def initialize(options = {})
+    # The {Writer} is also automatically started when necessary during calls to {.send_spans}.
+    #
+    # @public_api
     def start
```
Rephrase it.

marcotc commented on yard/extensions.rb:
```diff
@@ -1,5 +1,34 @@
 # frozen_string_literal: true
 
+TOP_LEVEL_MODULE_FILE = 'lib/ddtrace.rb'
+
+# The top-level `Datadog` module gets its docstring overwritten by
+# on almost every file in the repo, due to comments at the top of the file
+# (e.g. '# typed: true' or from vendor files 'Copyright (c) 2001-2010 Not Datadog.')
+#
+# This module ensures that only the comment provided by 'lib/ddtrace.rb'
+# is used as documentation for the top-level `Datadog` module.
+#
+# For non-top-level documentation, this can be solved by removing duplicate module/class
+# documentation. But for top-level it's tricky, as it is common to leave general comments
+# and directives in the first lines of a file.
+module EnsureTopLevelModuleCommentConsistency
+  def register_docstring(object, *args)
+    if object.is_a?(YARD::CodeObjects::ModuleObject) && object.path == 'Datadog' && parser.file != TOP_LEVEL_MODULE_FILE
+      super(object, nil)
+    else
+      super
+    end
+  end
+end
+YARD::Handlers::Base.prepend(EnsureTopLevelModuleCommentConsistency)
```
I agree that this should be super common.

I'll open an issue there, as I'm not super sure what the correct approach is.

marcotc commented on lib/ddtrace/sampler.rb:
```diff
@@ -6,22 +6,44 @@
     end
   end
 
   # \AllSampler samples all the traces.
```
Changed all `\Class` to `{Class}`.

marcotc commented:
I've addressed all the concerns that I was able to in this PR, @ivoanjo. Thank you so much for this monumental review 🙇.

As it stands, it's simple for us to add new methods to the Public API that I might have missed in this PR (that marking can be safely introduced in 1.0.1 or 1.1).
Better documentation is also something that can be always safely addressed.

Ivo helped me a lot to clean up public methods that we don't need to be exposed, as those are the ones I'm most concerned about, as removing them is a breaking change.

I'm going to review this PR once again with @delner in the new year, before we release 1.0, specially focusing on not exposing more than we need.

At this time, in order to unblock the 1.0 branch, I'm going to merge this PR, as I believe it is a good state.

ivoanjo commented:
> I'm going to review this PR once again with @delner in the new year, before we release 1.0, specially focusing on not exposing more than we need.
>
> At this time, in order to unblock the 1.0 branch, I'm going to merge this PR, as I believe it is a good state.

Thanks for looking into it! Fully agree with you -- the biggest concern right now, beyond making the docs useful and usable to customers, is with marking as public something that shouldn't be; forgetting something is less important.
